aten/src/ATen/native/cuda/Loss.cu

```diff
@@ -153,4 +154,290 @@ Tensor& binary_cross_entropy_backward_out_cuda(const Tensor& grad, const Tensor&
+    accscalar_t output_acc = 0;
+    accscalar_t total_weight_acc = 0;
+    for (i = 0; i < NLL_LOSS_THREADS; ++i) {
+      output_acc += sh_inputs[i];
+      total_weight_acc += acc_weight[i];
+    }
+    *total_weight = static_cast<scalar_t>(total_weight_acc);
+    *output = static_cast<scalar_t>(output_acc);
+    if (size_average) {
+      if (nframe == 0) {
+        // Mean reduction on empty tensors produces NaN
+        *output = std::numeric_limits<double>::quiet_NaN();
+      }
+      if (*total_weight != 0) {
+        *output = static_cast<scalar_t>(output_acc / total_weight_acc);
+      }
+    }
+  }
+}
+
+} // namespace
+
+void nll_loss_forward_out_cuda_template(
+    Tensor& output,
+    Tensor& total_weight,
+    const Tensor& input,
+    const Tensor& target,
+    const c10::optional<Tensor>& weight_opt,
+    int64_t reduction,
+    int64_t ignore_index) {
+  // See [Note: hacky wrapper removal for optional tensor]
+  c10::MaybeOwned<Tensor> weight_maybe_owned =
+      at::borrow_from_optional_tensor(weight_opt);
+  const Tensor& weight = *weight_maybe_owned;
+
+  TORCH_CHECK(target.dim() <= 1, "multi-target not supported");
+
+  int64_t n_classes = input.size(-1);
+  int64_t n_dims = input.dim();
+
+  TensorArg output_arg{output, "output", 1};
+  TensorArg total_weight_arg{total_weight, "total_weight", 2};
+  TensorArg input_arg{input, "input", 3};
+  TensorArg target_arg{target, "target", 4};
+
+  if (weight.defined()) {
+    TensorArg weight_arg{weight, "weight", 5};
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda_template",
+        {output_arg, total_weight_arg, input_arg, target_arg, weight_arg});
+  } else {
+    checkAllSameGPU(
+        "nll_loss_forward_out_cuda_template",
+        {output_arg, total_weight_arg, input_arg, target_arg});
+  }
+
+  TORCH_CHECK(n_dims > 0 && n_dims <= 2, "input tensor should be 1D or 2D");
+  int64_t batch_size = n_dims == 1 ? 1 : input.size(0);
+  int64_t num_targets = target.size(0);
+  TORCH_CHECK(
+      batch_size == num_targets,
+      "size mismatch (got input: ",
+      input.sizes(),
+      ", target: ",
+      target.sizes(),
+      ")")
+
+  TORCH_CHECK(
+      !weight.defined() || weight.numel() == n_classes,
+      "weight tensor should be defined either for all ",
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  if (reduction == Reduction::None & n_dims == 2) {
+    output.resize_({batch_size});
+    if (batch_size == 0) {
+      // This guards from unnecessary operations and launching CUDA kernel with
+      // 0 blocks.
+      return;
+    }
+
+    AT_DISPATCH_FLOATING_TYPES_AND2(
+        at::ScalarType::Half,
+        at::ScalarType::BFloat16,
+        input.scalar_type(),
+        "nll_loss_forward_no_reduce_cuda_kernel",
+        [&] {
+          scalar_t* weight_data = nullptr;
+          if (weight.defined()) {
+            auto weight_ = weight.contiguous();
```

Comment:
`contiguous` should be called outside of dispatch guard, and here you can just set `weight_data` with a ternary operator